riscv/src/reachability.rs

```diff
@@ -137,22 +142,34 @@ pub fn extract_label_offsets(statements: &[Statement]) -> BTreeMap<&str, usize>
 }
 
 pub fn references_in_statement(statement: &Statement) -> BTreeSet<&str> {
+    let mut ret = BTreeSet::new();
     match statement {
-        Statement::Label(_) | Statement::Directive(_, _) => Default::default(),
-        Statement::Instruction(_, args) => args
-            .iter()
-            .filter_map(|arg| match arg {
-                Argument::Register(_) | Argument::StringLiteral(_) => None,
-                Argument::Symbol(s) => Some(s.as_str()),
-                Argument::RegOffset(_, c) | Argument::Constant(c) => match c {
-                    Constant::Number(_) => None,
-                    Constant::HiDataRef(s, _offset) | Constant::LoDataRef(s, _offset) => {
-                        Some(s.as_str())
+        Statement::Label(_) | Statement::Directive(_, _) => (),
+        Statement::Instruction(_, args) => {
+            for arg in args {
+                match arg {
+                    Argument::Expression(expr) | Argument::RegOffset(_, expr) => {
+                        references_in_expressions(expr, &mut ret);
                     }
-                },
-                Argument::Difference(_, _) => todo!(),
-            })
-            .collect(),
+                    _ => (),
+                }
+            }
+        }
+    };
+    ret
+}
+
+fn references_in_expressions<'a>(expression: &'a Expression, symbols: &mut BTreeSet<&'a str>) {
```

Comment:
Can you write a visitor like https://github.com/powdr-org/powdr/pull/311/files#diff-fa50f194bff4a9264f97a398686cfc32a2cc38f258b3320147718d091df7fe86R191 ? I think you can make use of it in multiple places.

Reply:
Done. I find I hate rust not being generic over `mut`.